code/game/machinery/shieldgen.dm

```diff
@@ -305,142 +305,115 @@
 
 	if(!PN)
-		power = FALSE
-		return
+		activated = FALSE
+		return FALSE
 
-	var/surplus = max(PN.avail-PN.load, 0)
-	var/shieldload = min(rand(50,200), surplus)
-	if(shieldload == 0 && !storedpower)		// no cable or no power, and no power stored
-		power = FALSE
-		return
-	else
-		power = TRUE	// IVE GOT THE POWER!
-		if(PN) //runtime errors fixer. They were caused by PN.load trying to access missing network in case of working on stored power.
-			storedpower += shieldload
-			PN.load += shieldload //uses powernet power.
-//		message_admins("[PN.load]", 1)
-//		use_power(250) //uses APC power
+	var/surplus = max(PN.avail - PN.load, 0)
+	var/shieldload = min(rand(50, 200), surplus)
+	if(!shieldload && stored_power <= 0)		// no cable or no power, and no power stored
+		activated = FALSE
+		return FALSE
+
+	stored_power += min(shieldload, MAX_STORED_POWER - stored_power)
+	PN.load += shieldload //uses powernet power.
+	return TRUE
 
 /obj/machinery/shieldwallgen/attack_hand(mob/user)
-	if(state != 1)
+	if(!anchored)
 		to_chat(user, "<span class='warning'>The shield generator needs to be firmly secured to the floor first.</span>")
 		return TRUE
 	if(locked && !issilicon(user))
 		to_chat(user, "<span class='warning'>The controls are locked!</span>")
 		return TRUE
-	if(!power)
+	var/turf/T = loc
+	if(!T.get_cable_node())
 		to_chat(user, "<span class='warning'>The shield generator needs to be powered by wire underneath.</span>")
 		return TRUE
 
-	active = !active
-	user.visible_message("[user] turned the shield generator [active ? "on" : "off"].", \
-	"You turn [active ? "on" : "off"] the shield generator.", \
-	"You hear heavy droning [active ? "" : "fade out"].")
-	if(active)
-		for(var/dir in list(NORTH, SOUTH, EAST, WEST))
-			cleanup(dir)
+	if(!activated)
+		activate()
+	else
+		deactivate()
+
+	user.visible_message("<span class='notice'>[user] turned the shield generator [activated ? "on" : "off"].</span>", \
+		"<span class='notice'>You turn [activated ? "on" : "off"] the shield generator.</span>", \
+		"You hear heavy droning [activated ? "" : "fade out"].")
+
 	update_icon(UPDATE_ICON_STATE)
 	add_fingerprint(user)
 
 /obj/machinery/shieldwallgen/process()
-	spawn(100)
-		power()
-		if(power)
-			storedpower -= 50 //this way it can survive longer and survive at all
-	if(storedpower >= maxstoredpower)
-		storedpower = maxstoredpower
-	if(storedpower <= 0)
-		storedpower = 0
-
-	if(active)
-		if(state)
-			active = FALSE
-			return
-		spawn(1)
-			setup_field(1)
-		spawn(2)
-			setup_field(2)
-		spawn(3)
-			setup_field(4)
-		spawn(4)
-			setup_field(8)
-	if(active)
-		if(!power)
-			visible_message("<span class='warning'>[name] shuts down due to lack of power!</span>", \
+	if(!activated)
+		return
+	if(!try_charge_shields_power())
+		visible_message("<span class='warning'>[name] shuts down due to lack of power!</span>", \
 				"You hear heavy droning fade out")
```

Comment:
I think the second arg is the message sent to the src. when this looks like it wants to be a blind message?
might want to use a named arg here.

Reply:
There are two versions of `visible_message`. One on mobs and one on everything else. The mobs one will take a `self_message` param on the second place. The base (atom) one will only take the message and the blind message.